File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.py
# 입력: N, K (수빈이의 위치, 동생의 위치)
# 로직: BFS를 이용해서 동생과 만날 때까지 수빈이가 이동할 수 있는 모든 경로를 이동
# 출력: 수빈이가 동생을 찾아가는 최소 시간

import sys
from collections import deque

input = sys.stdin.readline
n, k = map(int, input().split())

# bfs
def bfs(start, target):
  visited = [False] * 100_101
  queue = deque([[start, 0]])
  visited[start] = True
  
  while len(queue) > 0:
    current, time = queue.popleft()
    
    if current == target:
      return time
    
    if valid_range(current + 1) and visited[current + 1] == False:
      queue.append([current + 1, time + 1])
      visited[current + 1] = True
    if valid_range(current - 1) and visited[current - 1] == False:
      queue.append([current - 1, time + 1])
      visited[current - 1] = True
    if valid_range(current * 2) and visited[current * 2] == False:
      queue.append([current * 2, time + 1])
      visited[current * 2] = True
    
def valid_range(value):
  return value >= 0 and value <= 100_000

# 최소 시간 출력
print(bfs(n, k))

File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.js
const fs = require('fs');
const [n, k] = fs.readFileSync('/dev/stdin').toString().trim().split(' ').map(Number);
const time = Array.from({length: 100001}, () => -1);

function solution() {
    const queue = [n];
    time[n] = 0;
    while(queue.length > 0) {
        const x = queue.shift();
        if(x - 1 >= 0 && time[x - 1] < 0) {
            time[x - 1] = time[x] + 1;
            queue.push(x - 1);
        }
        if(x + 1 <= 100000 && time[x + 1] < 0) {
            time[x + 1] = time[x] + 1;
            queue.push(x + 1);
        }
        if(x * 2 <= 100000 && time[x * 2] < 0) {
            time[x * 2] = time[x] + 1;
            queue.push(x * 2);
        }
    }
    
    console.log(time[k]);
}

solution();